### ISSUE TITLE
DeltaTorch Compatability?

### ISSUE THREAD

**Issue by rangi513** [mosaicml/streaming#690]
I'm in the system design phase of a new project and due to the scale of our data we plan to run multi node + multi gpu training with Pytorch on Databricks. I'm particularly interested in using this library for the true determinism and instant mid-epoch resumption features highlighted in the readme.  

We store our training data in Delta tables (Lakehouse) today so we were planning on using [DeltaTorch](https://github.com/delta-incubator/deltatorch) which according to their release post:
> enables PyTorch DataLoaders to seamlessly use Delta tables as a data source. These DataLoaders handle the batching, sampling, and multiprocessing of data during PyTorch training pipelines, without requiring a temporary copy of the Delta tables

I'm curious if Mosaic StreamingDataset plays well with DeltaTorch and reading directly from this delta format without creating temporary copies. Or does the MDS format required by StreamingDataset make DeltaTorch useless? I'm trying to understand the overlap of these two technologies in the Databricks ecosystem to reduce our spark write and rewrite compute and be efficient about our distribution of training data to the various gpus on our nodes during training. Any guidance would be appreciated!

**Comment by karan6181** [comment]
Hi @rangi513, thank you for looking into the Streaming Dataset. The Streaming Dataset provides an API called [dataframe_to_mds()](https://github.com/mosaicml/streaming/blob/main/streaming/base/converters/dataframe_to_mds.py#L174) which takes a spark DataFrame and converts to an MDS format. You can use the DataBricks spark based multiprocessing to convert your dataset into an MDS format faster. Please checkout one of the examples [here](https://docs.mosaicml.com/projects/streaming/en/stable/preparing_datasets/spark_dataframe_to_mds.html).

With regards to your dataset:
- How big is your dataset?
- What are your dataset columns or fields?
- Are you looking for any dataset filteration on the fly?

**Comment by rangi513** [comment]
100-500 million user events per model (creating a handful of different models here predicting the same set of targets MTL). I will do spark preprocessing on the raw event data (spark df) to produce multi dimensional numpy arrays/numeric tensors usable in training by pytorch. Filtration will be done ahead of time as well during the spark preprocessing so it doesn't necessarily need to be done in the dataloader class (although that can be a plus if its easy to implement with this).

I'm trying to understand whether it's going to be more compute+memory efficient to take a spark dataframe and write it to MDS format and train with Streaming Dataset or write it to Delta format and use DeltaTorch to load it directly into PyTorch without creating temp spark copies. Is there any overlap between these two Databricks technologies for training with Pytorch or are they generally expected to be used mutually exclusively?

**Comment by karan6181** [comment]
Thank you for the information. Currently, DeltaTorch and Streaming Dataset do not overlap. However, for your use case, filtering during data pre-processing with Spark and saving in MDS format can be very efficient. Using Streaming Dataset with MDS format is compute and memory efficient due to its low memory usage per sample and high performance for single sample queries. Additionally, you will benefit from features like deterministic shuffling, elastic resumption, multiple dataset mixing, dynamic batching composition, and shard eviction on the fly.
